File: src/main/java/nl/sourcelabs/demos/cf/productdata/ProductDataService.java
package nl.sourcelabs.demos.cf.productdata;

import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import nl.sourcelabs.demos.cf.domain.Price;
import nl.sourcelabs.demos.cf.domain.Product;
import nl.sourcelabs.demos.cf.domain.Review;
import nl.sourcelabs.demos.cf.domain.Stock;
import org.springframework.boot.web.client.RestTemplateBuilder;
import org.springframework.stereotype.Service;
import org.springframework.util.StopWatch;
import org.springframework.web.client.RestTemplate;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

@Service
@Slf4j
public class ProductDataService {

    private RestTemplate restTemplate;

    public ProductDataService(RestTemplateBuilder builder) {
        restTemplate = builder
                .rootUri("http://localhost:1080")
                .build();
    }

    /***
     * 1. /products list products
     * 2. /prices?productId={} price for product (May not produce error or no data)
     * 3. /warehouses?productId={} returns locationId for stock query (can product default or error)
     * 4. /locations?locationId={}&productId={} returns stock for product (can produce default on error)
     * 5. /reviews?productId={} list of reviews
     * @return
     */
    public ProductResponse getProducts() {
        StopWatch sp = new StopWatch();
        sp.start();
        try {
            return new ProductResponse(
                    getProductList().getProducts().stream()
                            .map(this::addData)
                            .collect(Collectors.toList())
            );
        } finally {
            sp.stop();
            log.info("Duration:" + sp.getTotalTimeMillis() + "ms");
        }

    }

    private Product addData(Product initialProduct) {
        setProductReviews(initialProduct);
        setProductPrice(initialProduct);
        setProductStock(initialProduct);
        return initialProduct;
    }

    private void setProductReviews(Product initialProduct) {
        try {
            initialProduct.setReviews(
                    getProductReviews(initialProduct.getId()).stream()
                            .map(rpr -> Review.builder()
                                    .stars(Integer.valueOf(rpr.stars))
                                    .comment(rpr.descr)
                                    .build())
                            .collect(Collectors.toList()));
        } catch (Exception e) {
            log.error("oops, review is broken for product {}", initialProduct.getId());
        }
    }

    private void setProductPrice(Product initialProduct) {
        try {
            RemoteProductPrices rpp = getProductPrice(initialProduct.getId());
            initialProduct.setPrice(Price.builder()
                    .amount(rpp.price)
                    .currency(rpp.getCurrency())
                    .build());
        } catch (Exception e) {
            log.error("oops, price is broken for product {}", initialProduct.getId());
        }
    }

    private void setProductStock(Product initialProduct) {
        try {
            initialProduct.setStock(getProductStock(initialProduct.getId()));
        } catch (Exception e) {
            log.error("oops, stock is broken for product {}", initialProduct.getId());
        }
    }

    // First call to get all products
    private RemoteProductResponse getProductList() {
        return restTemplate.getForObject("/products", RemoteProductResponse.class);
    }

    // Get price for a product
    private RemoteProductPrices getProductPrice(String productId) {
        return restTemplate.getForObject("/prices?productId=" + productId, RemoteProductPrices.class);
    }

    // Get reviews for a product
    private List<RemoteProductReview> getProductReviews(String productId) {
        return Arrays.asList(restTemplate.getForObject("/reviews?productId=" + productId, RemoteProductReview[].class));
    }

    // Get stock for a product
    private Stock getProductStock(String productId) {
        RemoteProductWarehouse rpw = restTemplate
                .getForObject("/warehouses?productId=" + productId, RemoteProductWarehouse.class);
        RemoteProductStock rps = restTemplate
                .getForObject("/locations?productId=" + productId + "&locationId=" + rpw.getLocationId(),
                        RemoteProductStock.class);
        return Stock.builder()
                .location(rpw.getName())
                .level(Integer.valueOf(rps.getStockCount()))
                .build();
    }

    @Data
    @NoArgsConstructor
    private static class RemoteProductResponse {

        private List<Product> products;
    }

    @Data
    @NoArgsConstructor
    private static class RemoteProductPrices {

        private String price;
        private String currency;
    }

    @Data
    @NoArgsConstructor
    private static class RemoteProductReview {

        private String descr;
        private String stars;
    }

    @Data
    @NoArgsConstructor
    private static class RemoteProductWarehouse {

        private String warehouseId;
        private String name;
        private String locationId;
    }

    @Data
    @NoArgsConstructor
    private static class RemoteProductStock {

        private String stockCount;
    }

}

File: src/main/java/nl/sourcelabs/demos/cf/productdata/ProductDataService.kt
package nl.sourcelabs.demos.cf.productdata

import kotlinx.coroutines.*
import nl.sourcelabs.demos.cf.domain.*
import org.springframework.stereotype.Service
import org.springframework.web.client.RestTemplate
import org.springframework.web.client.getForObject

@Service
class ProductDataService(
    private val client: Client
) {

    suspend fun getProducts() = withContext(Dispatchers.IO) {
        ProductResponse(
            products = client.loadProducts()
                .map { async { enrich(it) } }
                .awaitAll()
                .filter { it.price != null }
        )
    }

    private suspend fun enrich(product: Product) = coroutineScope {
        val price = async {
            client.loadPrices(product)?.let { Price(amount = it.price, currency = it.currency) }
        }

        val stock = async {
            client.loadWarehouses(product)?.let {
                val remoteStock = client.loadStock(product, it.locationId!!)
                Stock(level = remoteStock?.stockCount?.toInt() ?: 0, location = it.name)
            }
        }

        val reviews = async {
            client.loadReviews(product).map {
                Review(stars = it.stars ?: 0, comment = it.descr ?: "n/a")
            }
        }

        product.copy(
            price = price.await(),
            stock = stock.await(),
            reviews = reviews.await()
        )
    }
}

@Service
class Client(private val restTemplate: RestTemplate) {

    suspend fun loadProducts(): List<Product> = try {
        restTemplate.getForObject<RemoteProductResponse>("/products").products
    } catch (exception: Exception) {
        emptyList()
    }

    suspend fun loadPrices(product: Product): RemoteProductPrice? = try {
        restTemplate.getForObject("/prices?productId=${product.id}")
    } catch (exception: Exception) {
        null
    }

    suspend fun loadWarehouses(product: Product): RemoteProductWarehouse? = try {
        restTemplate.getForObject("/warehouses?productId=${product.id}")
    } catch (exception: Exception) {
        null
    }

    suspend fun loadStock(product: Product, locationId: String): RemoteProductStock? = try {
        restTemplate.getForObject("/locations?productId=${product.id}&locationId=${locationId}")
    } catch (exception: Exception) {
        null
    }

    suspend fun loadReviews(product: Product): List<RemoteProductReview> = try {
        restTemplate.getForObject<Array<RemoteProductReview>>("/reviews?productId=${product.id}").toList()
    } catch (exception: Exception) {
        emptyList()
    }
}